### PR TITLE
[BEAM-11196] Set parent of fused stages to the lowest common ancestor

### PR DESCRIPTION
**Pull request by yifanmai**
Currently the parent of the fused stage is set to `None` if the two stages do not share a common parent. This sets it to the lowest common ancestor, if one exists.

------------------------

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Dataflow | Flink | Samza | Spark | Twister2
--- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/) | ---
Java | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/)
Python | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/) | ---
XLang | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/) | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website | Whitespace | Typescript
--- | --- | --- | --- | --- | --- | ---
Non-portable | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/) <br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/)
Portable | --- | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | --- | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by yifanmai** [comment]
R: @robertwb

**Comment by yifanmai** [comment]
Flink is still unhappy with this change. I will investigate further.

**Comment by yifanmai** [comment]
R: @robertwb @tvalentyn 

Ready for review. PTAL, thanks!

**Comment by tvalentyn** [comment]
What is the context of this change?

**Comment by yifanmai** [comment]
Summarizing an an offline conversation with @robertwb: If you fuse two stages without a common parent, currently the parent gets set to `None` i.e. it gets promoted to a root transform. This is problematic because in [some places](https://github.com/apache/beam/blob/eeb1497875bd9f9ae1be60e4b8d2d05dd834064e/sdks/python/apache_beam/pipeline.py#L892) we assume that there is exactly one root transform. Instead we should set the parent to the lowest common ancestor.

The change also applies this logic to `eliminate_common_key_with_void` and `pack_combiners`, though these are pseudo-fusions (i.e. replacing multiple stages with a single stage) rather than a "true" fusion.

**Comment by tvalentyn** [comment]
> Summarizing an an offline conversation with @robertwb: If you fuse two stages without a common parent, currently the parent gets set to None i.e. it gets promoted to a root transform. This is problematic because in some places we assume that there is exactly one root transform. Instead we should set the parent to the lowest common ancestor.


Thanks, let's add this to a JIRA and tag the issue ID in PR description.

File: sdks/python/apache_beam/runners/portability/fn_api_runner/translations.py [hunk 1]
```diff
@@ -1166,6 +1173,36 @@ def unlifted_stages(stage):
...
+  # type: (str, str, TransformContext) -> Optional[str]
+
+  '''Returns the name of the lowest common ancestor of the two named stages.
+
+  The provided context is used to compute ancestors of stages. Note that stages
+  are considered to be ancestors of themselves.
+  '''
+  assert a != b
+
+  parents = {
```
Comment (tvalentyn) [inline sdks/python/apache_beam/runners/portability/fn_api_runner/translations.py:1186]: Consider following for easier readability:
```
  parents = {
      child_id: parent_id
      for parent_id, parent in context.components.transforms.items()
      for child_id in parent.subtransforms
  }
```

**Comment by tvalentyn** [comment]
Is there a risk that this will break streaming job updates on affected pipelines?

**Comment by robertwb** [comment]
This will not break streaming job updates because the translation to DFE ignores parents (aka composite transforms). It also doesn't (yet) use these codepaths. 

FWIW, the other reason this is problematic is that nulling out the parent can create cases where A -> B -> C and A and C are inside the composite but B is not.

File: sdks/python/apache_beam/runners/portability/fn_api_runner/translations.py (hunk 1 above)
Comment (robertwb) [reply sdks/python/apache_beam/runners/portability/fn_api_runner/translations.py:1186]: Yeah, this is an issue with yapf.

**Review by robertwb — APPROVED** [review]

**Comment by yifanmai** [comment]
Thanks for the fixes @robertwb !

**Pull request merged**